### PR TITLE
Modify Hadoop exploiter

### LINKED ISSUE #1605: Implement exploitation in Puppet
Implement the `exploit_host()` function of the concrete Puppet. Modify the Master to pass the appropriate options to the exploiters (i.e. propagation credentials). The `smb` exploit may require additional options.

**Note**
There's currently a bug where MSSQL exploiter can fail to propagate and prevent other exploiters from running. We can most likely resolve this by having each exploiter return two booleans, one to indicate whether or not the victim is vulnerable and another to indicate whether or not propagation was successful. This would also let us simplify the exploiters config, which now have a true/false field to indicate whether or not an exploiter is capable of propagating.

## Tasks

- [x] Modify config to remove boolean "propagator" field from config (0d) @ilija-lazoroski 
- [x] Add a periodic check (between hosts and/or between exploiters) to get updated credentials from the Island (0d) @mssalvatore 
- [x] Pass other WormConfig settings as options (0d) @ilija-lazoroski 
    - [x] Target dropper paths
    - [x] Pass credentials as options to exploiters (See #1721 for a starting point)
- [x] Copy commonly used functions on WormConfig to a utils/tools module and add tests (0d) @mssalvatore
    - [x] get_exploit_user_password_or_hash_product()
    - [x] get_exploit_user_ssh_key_pairs()
    - [x] get_exploit_user_password_pairs()
- [x] Modify ExploitResultData (0d) @shreyamalviya
    - [x] Add OS
    - [x] Add `exploit_successful`
    - [x] Add `propagation_successful`
    - [x] Modify mock_puppet
- [x] Remove `skip_if_exploit_file_exists` option from config (0d) @shreyamalviya
- [x] SSH (0d) @ilija-lazoroski 
    - [x] Don't modify VictimHost object
    - [x] Change attempt logging and remove credential hashes from log statements
    - [x] Return ExploitResultData
    - [x] Call new `get_exploit_user*()` functions instead of `self._config.*()`
    - [x] Accept ITelemetryMessenger in `__init__()` and use it to send telemetries
- [x] Hadoop (0d) @shreyamalviya 
    - [x] Remove code that sets architecture from web_rce.py
    - [x] Return ExploitResultData
    - [x] Accept ITelemetryMessenger in `__init__()` and use it to send telemetries
    - [x] Wrap ~Struts~Hadoop with wrapper and load into the puppet in `monkey.py`
- [x] Remove dependency on `Plugin` from `HostExploiter` (0d) @mssalvatore 
- [x] Implement `exploit_host()` in `Puppet` (0d) @mssalvatore 
    - [x] Add a wrapper around Exploiters that accepts the necessary parameters for exploitation and object construction (VictimHost, Telemetry messenger, options)

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #1605 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] ~Is the TravisCI build passing?~
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

image
image
image
image

### CODE REVIEW

**Review by ilija-lazoroski — APPROVED** [review]
LGTM!

File: monkey/infection_monkey/exploit/hadoop.py [hunk 1]
```diff
@@ -33,29 +33,28 @@ class HadoopExploiter(WebRCE):
...
-            return False
+            return self.exploit_result
         http_path, http_thread = HTTPTools.create_locked_transfer(self.host, paths["src_path"])
-        command = self.build_command(paths["dest_path"], http_path)
+        command = self._build_command(paths["dest_path"], http_path)
         if not self.exploit(self.vulnerable_urls[0], command):
-            return False
+            return self.exploit_result
         http_thread.join(self.DOWNLOAD_TIMEOUT)
         http_thread.stop()
```
Comment (VakarisZ) [inline monkey/infection_monkey/exploit/hadoop.py:53]: Hmmm, shouldn't this happen regardless of success/failure?

Comment (mssalvatore) [reply monkey/infection_monkey/exploit/hadoop.py:53]: Yeah. Let's wrap it in a `try/finally`.

Comment (mssalvatore) [reply monkey/infection_monkey/exploit/hadoop.py:53]: Yeah. I recommend something like this:

```python
        paths = self.get_monkey_paths()
        if not paths:
            return self.exploit_result

        http_path, http_thread = HTTPTools.create_locked_transfer(self.host, paths["src_path"])

        try:
            command = self._build_command(paths["dest_path"], http_path)

            if self.exploit(self.vulnerable_urls[0], command):
                self.add_executed_cmd(command)
                self.exploit_result.exploitation_success = True
                self.exploit_result.propagation_success = True
        finally:
            http_thread.join(self.DOWNLOAD_TIMEOUT)
            http_thread.stop()

        return self.exploit_result
```

**Pull request merged**